File: scripts/differential/java/P2T13.java
package app.freerouting.datastructures;

import app.freerouting.geometry.planar.IntPoint;
import app.freerouting.geometry.planar.Point;
import java.util.Collection;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

/**
 * Java ground truth for {@link PlanarDelaunayTriangulation} (Plan 2 Task 13). Twin: `p2t13`.
 *
 * <p>args: n seed mode
 *
 * <p>Modes:
 *
 * <ul>
 *   <li>0 — n random points, one single-corner object each, coordinates in [-100000, 100000).
 *   <li>1 — the four corners of a square (the brief's pinned case).
 *   <li>2 — a collinear triple.
 *   <li>3 — a square plus three coincident interior points (duplicate handling).
 *   <li>4 — n objects of *two* corners each, so `corner.object == commonCorner.object` (the
 *       same-object duplicate short-circuit at PlanarDelaunayTriangulation.java:157) is reachable.
 *   <li>5 — a 5x5 integer grid: maximal cocircularity and collinearity.
 *   <li>6 — n random points in the tiny range [-20, 20), which forces duplicates and collinear
 *       triples at a high rate.
 *   <li>7 — n random points on a circle of radius 30000 (every point cocircular).
 * </ul>
 *
 * <p>Prints the shuffle permutation the class's fixed-seed `Collections.shuffle` produces for the
 * corner count in play (re-derived here with an identical `LinkedList` + `new Random(99)`, because
 * the real one is private), then `getEdgeLines()` in iteration order, then `validate()`.
 */
public class P2T13 {

  static long state;

  static long next() {
    state ^= state << 13;
    state ^= state >>> 7;
    state ^= state << 17;
    return state;
  }

  static int rnd(int bound) {
    return (int) Long.remainderUnsigned(next(), bound);
  }

  static class Obj implements PlanarDelaunayTriangulation.Storable {

    final int id;
    final Point[] corners;

    Obj(int id, Point... corners) {
      this.id = id;
      this.corners = corners;
    }

    @Override
    public Point[] getTriangulationCorners() {
      return corners;
    }
  }

  static Point p(int x, int y) {
    return new IntPoint(x, y);
  }

  static String fmt(Point point) {
    if (point == null) {
      return "null";
    }
    IntPoint ip = (IntPoint) point;
    return ip.x + "," + ip.y;
  }

  static String objId(PlanarDelaunayTriangulation.Storable object) {
    return object == null ? "none" : Integer.toString(((Obj) object).id);
  }

  public static void main(String[] args) {
    int n = args.length > 0 ? Integer.parseInt(args[0]) : 50;
    long seed = args.length > 1 ? Long.parseLong(args[1]) : 42;
    int mode = args.length > 2 ? Integer.parseInt(args[2]) : 0;
    state = seed;

    List<PlanarDelaunayTriangulation.Storable> objects = new LinkedList<>();
    int nextId = 1;
    if (mode == 1) {
      objects.add(new Obj(nextId++, p(0, 0)));
      objects.add(new Obj(nextId++, p(1000, 0)));
      objects.add(new Obj(nextId++, p(1000, 1000)));
      objects.add(new Obj(nextId++, p(0, 1000)));
    } else if (mode == 2) {
      objects.add(new Obj(nextId++, p(0, 0)));
      objects.add(new Obj(nextId++, p(500, 500)));
      objects.add(new Obj(nextId++, p(1000, 1000)));
    } else if (mode == 3) {
      objects.add(new Obj(nextId++, p(0, 0)));
      objects.add(new Obj(nextId++, p(1000, 0)));
      objects.add(new Obj(nextId++, p(1000, 1000)));
      objects.add(new Obj(nextId++, p(0, 1000)));
      objects.add(new Obj(nextId++, p(300, 400)));
      objects.add(new Obj(nextId++, p(300, 400)));
      objects.add(new Obj(nextId++, p(300, 400)));
    } else if (mode == 4) {
      for (int i = 0; i < n; i++) {
        objects.add(
            new Obj(
                nextId++,
                p(rnd(200000) - 100000, rnd(200000) - 100000),
                p(rnd(200000) - 100000, rnd(200000) - 100000)));
      }
    } else if (mode == 5) {
      for (int x = 0; x < 5; x++) {
        for (int y = 0; y < 5; y++) {
          objects.add(new Obj(nextId++, p(x * 1000, y * 1000)));
        }
      }
    } else if (mode == 6) {
      for (int i = 0; i < n; i++) {
        objects.add(new Obj(nextId++, p(rnd(40) - 20, rnd(40) - 20)));
      }
    } else if (mode == 7) {
      for (int i = 0; i < n; i++) {
        double angle = 2.0 * Math.PI * i / n;
        objects.add(
            new Obj(
                nextId++,
                p((int) Math.round(30000.0 * Math.cos(angle)),
                    (int) Math.round(30000.0 * Math.sin(angle)))));
      }
    } else {
      for (int i = 0; i < n; i++) {
        objects.add(new Obj(nextId++, p(rnd(200000) - 100000, rnd(200000) - 100000)));
      }
    }

    // Re-derive the permutation the constructor's private shuffle produces, so the Rust twin can
    // be checked against it directly rather than only through the triangulation result.
    int cornerCount = 0;
    StringBuilder inputs = new StringBuilder();
    for (PlanarDelaunayTriangulation.Storable object : objects) {
      for (Point corner : object.getTriangulationCorners()) {
        inputs.append(" ").append(objId(object)).append("@").append(fmt(corner));
        cornerCount++;
      }
    }
    System.out.println("mode=" + mode + " objects=" + objects.size() + " corners=" + cornerCount);
    System.out.println("in:" + inputs);

    List<Integer> permutation = new LinkedList<>();
    for (int i = 0; i < cornerCount; i++) {
      permutation.add(i);
    }
    Random shuffler = new Random(99);
    shuffler.setSeed(99);
    Collections.shuffle(permutation, shuffler);
    System.out.println("perm=" + permutation);

    PlanarDelaunayTriangulation triangulation = new PlanarDelaunayTriangulation(objects);
    Collection<PlanarDelaunayTriangulation.ResultEdge> edges = triangulation.getEdgeLines();
    System.out.println("count=" + edges.size());
    int i = 0;
    for (PlanarDelaunayTriangulation.ResultEdge edge : edges) {
      System.out.println(
          "E"
              + i
              + " so="
              + objId(edge.startObject)
              + " sp="
              + fmt(edge.startPoint)
              + " eo="
              + objId(edge.endObject)
              + " ep="
              + fmt(edge.endPoint));
      i++;
    }
    System.out.println("validate=" + triangulation.validate());
  }
}

File: scripts/differential/rust/src/bin/p2t13.rs
//! Rust twin of `scripts/differential/java/P2T13.java` (Plan 2 Task 13).
//!
//! Feeds identical point sets to `copper_board::datastructures::delaunay::PlanarDelaunayTriangulation`
//! and prints `getEdgeLines()` in iteration order, plus the shuffle permutation and `validate()`.
//! See `scripts/differential/README.md`.

use copper_board::datastructures::delaunay::{DelaunayCorner, PlanarDelaunayTriangulation};
use copper_board::ids::ItemId;
use copper_geometry::int_point::IntPoint;
use copper_geometry::point::Point;

static mut STATE: u64 = 0;

fn next() -> u64 {
    unsafe {
        STATE ^= STATE << 13;
        STATE ^= STATE >> 7;
        STATE ^= STATE << 17;
        STATE
    }
}

fn rnd(bound: u64) -> i32 {
    (next() % bound) as i32
}

fn p(x: i32, y: i32) -> Point {
    Point::Int(IntPoint::new(x, y))
}

fn fmt(point: &Point) -> String {
    match point {
        Point::Int(ip) => format!("{},{}", ip.x, ip.y),
        Point::Rational(_) => "rational".to_string(),
    }
}

fn obj_id(object: Option<ItemId>) -> String {
    match object {
        None => "none".to_string(),
        Some(id) => id.0.to_string(),
    }
}

/// `java.util.Random` + `Collections.shuffle`, duplicated here (the ones inside `copper-board` are
/// private) so the driver can print the permutation the same way the Java driver does.
struct JavaRandom {
    seed: i64,
}

impl JavaRandom {
    fn new(seed: i64) -> JavaRandom {
        JavaRandom {
            seed: (seed ^ 0x5DEECE66D_i64) & ((1 << 48) - 1),
        }
    }

    fn next(&mut self, bits: u32) -> i32 {
        self.seed = self
            .seed
            .wrapping_mul(0x5DEECE66D_i64)
            .wrapping_add(0xB)
            & ((1 << 48) - 1);
        (self.seed >> (48 - bits)) as i32
    }

    fn next_int(&mut self, bound: i32) -> i32 {
        let mut r = self.next(31);
        let m = bound - 1;
        if bound & m == 0 {
            r = ((i64::from(bound) * i64::from(r)) >> 31) as i32;
        } else {
            let mut u = r;
            loop {
                r = u % bound;
                if u.wrapping_sub(r).wrapping_add(m) >= 0 {
                    break;
                }
                u = self.next(31);
            }
        }
        r
    }
}

fn permutation(n: usize) -> Vec<usize> {
    let mut list: Vec<usize> = (0..n).collect();
    let mut rng = JavaRandom::new(99);
    let mut i = list.len();
    while i > 1 {
        let j = rng.next_int(i as i32) as usize;
        list.swap(i - 1, j);
        i -= 1;
    }
    list
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let n: usize = args.first().map_or(50, |a| a.parse().unwrap());
    let seed: u64 = args.get(1).map_or(42, |a| a.parse().unwrap());
    let mode: i32 = args.get(2).map_or(0, |a| a.parse().unwrap());
    unsafe { STATE = seed };

    // One entry per Java `Storable`: its id and its corners.
    let mut objects: Vec<(u32, Vec<Point>)> = Vec::new();
    let mut next_id = 1u32;
    let mut add = |objects: &mut Vec<(u32, Vec<Point>)>, corners: Vec<Point>| {
        objects.push((next_id, corners));
        next_id += 1;
    };
    match mode {
        1 => {
            add(&mut objects, vec![p(0, 0)]);
            add(&mut objects, vec![p(1000, 0)]);
            add(&mut objects, vec![p(1000, 1000)]);
            add(&mut objects, vec![p(0, 1000)]);
        }
        2 => {
            add(&mut objects, vec![p(0, 0)]);
            add(&mut objects, vec![p(500, 500)]);
            add(&mut objects, vec![p(1000, 1000)]);
        }
        3 => {
            add(&mut objects, vec![p(0, 0)]);
            add(&mut objects, vec![p(1000, 0)]);
            add(&mut objects, vec![p(1000, 1000)]);
            add(&mut objects, vec![p(0, 1000)]);
            add(&mut objects, vec![p(300, 400)]);
            add(&mut objects, vec![p(300, 400)]);
            add(&mut objects, vec![p(300, 400)]);
        }
        4 => {
            for _ in 0..n {
                let a = p(rnd(200_000) - 100_000, rnd(200_000) - 100_000);
                let b = p(rnd(200_000) - 100_000, rnd(200_000) - 100_000);
                add(&mut objects, vec![a, b]);
            }
        }
        5 => {
            for x in 0..5 {
                for y in 0..5 {
                    add(&mut objects, vec![p(x * 1000, y * 1000)]);
                }
            }
        }
        6 => {
            for _ in 0..n {
                add(&mut objects, vec![p(rnd(40) - 20, rnd(40) - 20)]);
            }
        }
        7 => {
            for i in 0..n {
                let angle = 2.0 * std::f64::consts::PI * (i as f64) / (n as f64);
                add(
                    &mut objects,
                    vec![p(
                        (30_000.0 * angle.cos()).round() as i32,
                        (30_000.0 * angle.sin()).round() as i32,
                    )],
                );
            }
        }
        _ => {
            for _ in 0..n {
                add(
                    &mut objects,
                    vec![p(rnd(200_000) - 100_000, rnd(200_000) - 100_000)],
                );
            }
        }
    }

    let mut corners: Vec<DelaunayCorner> = Vec::new();
    let mut inputs = String::new();
    for (id, points) in &objects {
        for point in points {
            inputs.push_str(&format!(" {}@{}", id, fmt(point)));
            corners.push(DelaunayCorner::new(ItemId(*id), point.clone()));
        }
    }
    println!(
        "mode={} objects={} corners={}",
        mode,
        objects.len(),
        corners.len()
    );
    println!("in:{inputs}");
    let perm = permutation(corners.len());
    println!(
        "perm=[{}]",
        perm.iter()
            .map(|i| i.to_string())
            .collect::<Vec<_>>()
            .join(", ")
    );

    let triangulation = PlanarDelaunayTriangulation::new(&corners);
    let edges = triangulation.get_edge_lines();
    println!("count={}", edges.len());
    for (i, edge) in edges.iter().enumerate() {
        println!(
            "E{} so={} sp={} eo={} ep={}",
            i,
            obj_id(edge.start_object),
            fmt(&edge.start_point),
            obj_id(edge.end_object),
            fmt(&edge.end_point)
        );
    }
    println!("validate={}", triangulation.validate());
}
